<a href="https://colab.research.google.com/github/adavals/bedu-datascience-f2m2/blob/main/Postwork1_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Equipo 17

# POSTWORK 1 - DEFINICIÓN DEL PROBLEMA

#INTRODUCCIÓN

La canasta básica es un indicador económico importante que refleja el costo de los productos y servicios necesarios para satisfacer las necesidades básicas de una persona o familia. El conocimiento preciso y actualizado de los precios de la canasta básica es esencial para entender la inflación, la economía y la calidad de vida de la población. PROFECO, es una institución gubernamental encargada de proteger los derechos de los consumidores en México, que proporciona información sobre los precios de la canasta básica en forma de datos abiertos. Sin embargo, estos datos parecen no estar siendo utilizados adecuadamente por el público en general. Es necesario analizar y utilizar esta información de manera efectiva para tener una comprensión completa de la situación económica del país y para mejorar la calidad de vida de la población.

#IDENTIFICACIÓN DEL PROBLEMA:

1. Falta de publicación completa de los datos de la canasta básica en la revista del consumidor PROFECO.

2. No se encontraron artículos de internet que hicieran referencia a los datos abiertos.

3. La aplicación móvil "Quien es quien en los precios" es la única vía de comparación de precios disponible para el público en general.

4. Los datos abiertos publicados por PROFECO no suelen ser analizados o publicados de forma completa en medios accesibles para el público.

5. Falta de conocimiento del público en general sobre la disponibilidad y uso de los datos abiertos de la canasta básica.

#INVESTIGACIÓN

1. Revisar la información publicada en la página web oficial de PROFECO en México.
2. Consultar la Revista del Consumidor de PROFECO para ver si hay artículos que hagan referencia a la canasta básica y su análisis.
3. Verificar la disponibilidad y accesibilidad de la aplicación "Quien es Quien en los precios" para comparar precios y ver si se pueden utilizar para analizar los datos de la canasta básica.
4. Consultar las estadísticas nacionales y estudios económicos para obtener información relevante sobre la economía, la inflación y los precios en México.

#DEFINICION DEL PROBLEMA

En una economía en constante fluctuación, el aumento de precios de los productos de la canasta básica es una realidad que afecta a la mayor parte de la población. Muchos de nosotros nos enfrentamos a la difícil tarea de hacer que nuestro dinero alcance para cubrir las necesidades básicas, mientras los precios suben sin cesar. En este contexto, es fundamental tener herramientas que nos permitan comparar los precios de los productos de consumo diario y tomar decisiones de compra más informadas, con el objetivo de ahorrar en nuestros gastos. ¿Cómo podemos acceder a información precisa y confiable sobre los precios de los productos básicos en diferentes lugares de compra, para elegir el que nos ofrezca el mejor valor por nuestro dinero?

## Trabajos previos
- Se examinaron algunos contenidos de Revista del consumidor PROFECO y no se encontró una publicación que analizara o hiciera referencia a este tipo de datos abiertos en forma completa, sólo los artículos referentes a un tipo de producto en particular que se publican cada mes pueden estar utilizando la información de los datos abiertos.
- Se realizó búsqueda en internet de artículos que hicieran referencia a estos datos y no se encontró ninguno.
- Se encontró en el portal oficial del gobierno de México el sitio web de Quien es Quien en los precios, relacionado con uno de los archivos de datos abiertos publicados, pero el sitio parece estár desconectado de la información actual y no muestra datos (https://www.profeco.gob.mx/precios/canasta/default.aspx)
- Existe una aplicación en Google Playstore: Quien es quien en los precios, que ofrece algunas vías de comparación de precios y tiene una interfaz de usuario usable, aunque comprobamos que tiene fallas al iniciar en algunos dispositivos.

# POSTWORK 2 - PLANTEAMIENTO DE PREGUNTAS

De acuerdo con el problema definido, se han planteado las siguientes preguntas:  

*  ¿Cómo saber dónde comprar los productos de consumo?
*  ¿Qué tiendas tienen en general precios más bajos cerca de una ubicación en particular?
*  ¿Qué tienda tiene precios más bajos por tipo de producto (frutas, verduras, medicamentos, etc.)?
*  ¿Para un producto en particular, de interés de un consumidor, qué tienda lo ha vendido más barato?
*  ¿En qué estados los precios son más bajos/altos?
*  Para una ubicación en particular ¿en qué segmento se ubican los precios de las tiendas cercanas (menor, medio, mayor)?
*  ¿En qué tienda se puede conseguir el conjunto de los productos de la canasta básica a un precio más bajo?
*  ¿Qué tienda tiene precios más bajos/más altos en los diferentes estados de la República?









# POSTWORK 3 - COLECCIÓN DE DATOS

Tratando de contestar las preguntas del postwork 2, se buscaron datasets y APIs en diferentes sitios como Kaggle, gobierno, etc. pero queríamos que la información fuera actual y solo se encontraron en la Procuraduría Federal del Consumidor en México, organismo que publica, en forma abierta, los datos que recolecta sobre los precios de una gran cantidad de productos en diferentes supermercados, farmacias y tiendas de conveniencia. [Datos abiertos al consumidor en el sitio de PROFECO](https://datos.profeco.gob.mx/datos_abiertos/).

Estos datos están publicados en un archivo comprimido en formato .rar por año. Un archivo comprimido contiene varios archivos en formato excel csv. Cada archivo tiene en su nombre el numero de semana y el año en que fueron recolectados los datos, en general hay un archivo por semana, por lo que, para el público consumidor es difícil aprovecharlos.

El último archivo publicado de 2022 lo descargamos y está disponible en el siguiente enlace:
[Diciembre 2022](https://raw.githubusercontent.com/adavals/bedu-datascience-f2m2/main/datasets/532022.csv)

La descripción de la estructura del archivo está publicada bajo el título de [Diccionario de Datos](https://datos.profeco.gob.mx/diccionarioDatosQQP.php)

Tratamos de conectarnos con la [API de PROFECO](https://datos.gob.mx/busca/dataset/quien-es-quien-en-los-precios/resource/408f99a4-04a7-4726-806c-f28b2e80fe00) pero no estaba activa y no pudimos usarla. 

Para complementar la información de ubicación con el código postal se usó la API  [COPOMEX - Código Postal Mexicano](https://api.copomex.com/panel/).

La cual tras, previo registro se puede acceder al uso mediante un panel de control, donde podremos crear nuestro proyecto para que se nos genere la Token del API. También tendremos una guía de uso de esta API a nuestra disposición  en los enlaces de ayuda.


# POSTWORK 4 - PANDAS Y ANÁLISIS EXPLORATORIO DE DATOS

Al revisar los datos disponibles encontramos que para cada año hay 53 o 54 archivos csv, uno por semana. Para efecto de nuestro análisis vemos que usando uno de estos archivos, por ejemplo el más reciente, podemos abordar las preguntas que nos hemos planteado.

A continuación analizamos nuestro conjunto de datos usando pandas.



In [ ]:
#Cargamos la biblioteca pandas
import pandas as pd

In [ ]:
#Cargamos el archivo csv (que ha sido descargado en el github) como un dataframe y asignamos los nombres de las columnas
df = pd.read_csv('https://raw.githubusercontent.com/adavals/bedu-datascience-f2m2/main/datasets/532022.csv', names=['producto','presentacion','marca','categoria','grupo','precio','fecha','marca_tienda','tipo_tienda','nombre_tienda','direccion','estado','municipio','latitud','longitud'])

In [ ]:
#Por medio de la función head vemos los primeros registros
df.head(3)

producto             presentacion   marca  \
0   ACEITE    BOTELLA 1 LT. VEGETAL   1-2-3   
1   ACEITE  BOTELLA 850 ML. VEGETAL     AVE   
2   ACEITE   BOTELLA 946 ML. CANOLA  CANOIL   

                           categoria    grupo  precio       fecha  \
0  ACEITES Y GRASAS VEG. COMESTIBLES  BASICOS    56.0  2022-12-26   
1  ACEITES Y GRASAS VEG. COMESTIBLES    PACIC    43.9  2022-12-26   
2  ACEITES Y GRASAS VEG. COMESTIBLES  BASICOS    50.9  2022-12-26   

   marca_tienda                            tipo_tienda  \
0  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
1  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
2  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   

                          nombre_tienda  \
0  MEGA SORIANA SUCURSAL AGUASCALIENTES   
1  MEGA SORIANA SUCURSAL AGUASCALIENTES   
2  MEGA SORIANA SUCURSAL AGUASCALIENTES   

                                           direccion          estado  \
0  AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
1  AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
2  AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   

        municipio    latitud    longitud  
0  AGUASCALIENTES  21.916001 -102.290779  
1  AGUASCALIENTES  21.916001 -102.290779  
2  AGUASCALIENTES  21.916001 -102.290779

In [ ]:
#Por medio de la propiedad tail vemos los últimos registros
df.tail(3)

producto                         presentacion marca  \
256420              XL-3  CAJA CON 2 CARTERAS. 4 TABLETAS C/U   S/M   
256421           XL-3 VR                     CAJA 24 TABLETAS   S/M   
256422  TORTILLA DE MAIZ                         1 KG. GRANEL   S/M   

                             categoria         grupo  precio       fecha  \
256420                    MEDICAMENTOS  MEDICAMENTOS   18.97  2022-12-30   
256421                    MEDICAMENTOS  MEDICAMENTOS   88.91  2022-12-30   
256422  TORTILLAS Y DERIVADOS DEL MAIZ       BASICOS   20.00  2022-12-30   

                      marca_tienda                            tipo_tienda  \
256420                   ISSSTEZAC  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
256421                   ISSSTEZAC  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
256422  TORTILLERIAS TRADICIONALES                           TORTILLERÍAS   

               nombre_tienda  \
256420  ISSSTEZAC SUCURSAL 2   
256421  ISSSTEZAC SUCURSAL 2   
256422   TORTILLERIA LA CRUZ   

                                                direccion     estado  \
256420  BLVD LOPEZ PORTILLO 238, ENTRE CALLE SICOMORO ...  ZACATECAS   
256421  BLVD LOPEZ PORTILLO 238, ENTRE CALLE SICOMORO ...  ZACATECAS   
256422  DE LA CRUZ 9, ENTRE JOSE MARIA RODRIGUEZ Y COL...  ZACATECAS   

        municipio    latitud    longitud  
256420  ZACATECAS  22.757732 -102.538580  
256421  ZACATECAS  22.757732 -102.538580  
256422  ZACATECAS  22.750167 -102.520612

Hacemos el **análisis exploratorio** de datos para saber si:
* ¿El conjunto de datos que tengo realmente me sirve para responder algunas de las preguntas que me planteé?
     - Las preguntas que se plantearon incluyen clasificaciones de productos, precios de productos, comparaciones de precios, localizaciones de tiendas, etc y podemos ver que el conjunto de datos tiene direcciones y coordenadas geográficas de las tiendas que se pueden usar para saber dónde comprar, (incluso podemos obtener el Código postal de las direcciones para hacer consultas mediante este dato) categorías de productos, precios, etc; todos estos datos se pueden usar para obtener la respuesta de las preguntas que se plantearon al inicio del proyecto.

* ¿Qué tamaño tiene mi conjunto de datos? ¿Serán datos suficientes?

In [ ]:
#La función shape nos da información de la cantidad de renglones y columnas
df.shape

(256423, 15)

- El archivo  seleccionado corresponde a la última semana de diciembre de 2022 y contiene 256,423 registros con 15 columnas.



* ¿Qué columnas tengo y qué información tengo en cada una de esas columnas?

In [ ]:
#Por medio de la función dtypes vemos cada columna y su tipo de dato
df.dtypes

producto          object
presentacion      object
marca             object
categoria         object
grupo             object
precio           float64
fecha             object
marca_tienda      object
tipo_tienda       object
nombre_tienda     object
direccion         object
estado            object
municipio         object
latitud          float64
longitud         float64
dtype: object

|Columna|Descripción|
|---|---|
|producto|Nombre del producto.|
|presentacion|Descripción de la presentación del producto.
|marca|Descripción de la marca del producto.
|categoria|Nombre de la categoría a la cual pertenece el producto.
|grupo|Nombre del grupo al que pertenece el producto.
|precio|Precio del producto de venta al público.
|fecha|Día, mes y año en que fue capturado el precio del producto en el establecimiento.
|marca_tienda|Conjunto de establecimientos que son sucursales de una misma marca comercial.
|tipo_tienda|Descripción del giro comercial.
|nombre_tienda|Nombre comercial asociado a la razón social del establecimiento.
|direccion|Dirección del establecimiento.
|estado|Entidad federativa en la que se localiza el establecimiento.
|municipio|Nombre del municipio o demarcación territorial en el que se localiza el establecimiento.
|latitud|Localización del establecimiento con respecto al ecuador. Expresadas en grados decimales
|longitud|Localización del establecimiento con respecto al meridiano 0 . Expresadas en grados decimales

In [ ]:
#Mediante la función info vemos más información sobre las columnas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 256423 entries, 0 to 256422
Data columns (total 15 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   producto       256423 non-null  object 
 1   presentacion   256423 non-null  object 
 2   marca          256423 non-null  object 
 3   categoria      256423 non-null  object 
 4   grupo          256423 non-null  object 
 5   precio         256423 non-null  float64
 6   fecha          256423 non-null  object 
 7   marca_tienda   256423 non-null  object 
 8   tipo_tienda    256423 non-null  object 
 9   nombre_tienda  256423 non-null  object 
 10  direccion      256423 non-null  object 
 11  estado         256423 non-null  object 
 12  municipio      256423 non-null  object 
 13  latitud        255516 non-null  float64
 14  longitud       255516 non-null  float64
dtypes: float64(3), object(12)
memory usage: 29.3+ MB


* Los nombres que tienen mis columnas, ¿son el nombre más apropiado?

In [ ]:
#Podemos acceder a una lista de columnas usando la función columns
df.columns

Index(['producto', 'presentacion', 'marca', 'categoria', 'grupo', 'precio',
       'fecha', 'marca_tienda', 'tipo_tienda', 'nombre_tienda', 'direccion',
       'estado', 'municipio', 'latitud', 'longitud'],
      dtype='object')

- Los nombres de las columnas se asignaron al cargar el archivo csv de acuerdo a los datos observados y con base en el [Diccionario de Datos](https://datos.profeco.gob.mx/diccionarioDatosQQP.php) publicado por PROFECO.

* ¿Qué tipos de datos tengo en cada columna? ¿Parecen ser el tipo correcto de datos? ¿O es un tipo de datos "incorrecto"?



In [ ]:
#La función loc nos permite ver la información de una fila específica
df.loc[0]

producto                                                    ACEITE
presentacion                                 BOTELLA 1 LT. VEGETAL
marca                                                        1-2-3
categoria                        ACEITES Y GRASAS VEG. COMESTIBLES
grupo                                                      BASICOS
precio                                                        56.0
fecha                                                   2022-12-26
marca_tienda                                          MEGA SORIANA
tipo_tienda                  SUPERMERCADO / TIENDA DE AUTOSERVICIO
nombre_tienda                 MEGA SORIANA SUCURSAL AGUASCALIENTES
direccion        AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...
estado                                              AGUASCALIENTES
municipio                                           AGUASCALIENTES
latitud                                                  21.916001
longitud                                               -102.29

- El tipo de dato de cada uno es correcto a excepción de la fecha que presenta tipo object.
- En la columna ```presentacion``` se observa que los productos pueden venir en medidas diferentes como kilogramos, mililitros, litros, etc. por lo que se requiere hacer un tratamiento a la información para realizar comparaciones de precios correctas.


* Si selecciono algunas filas al azar y las observo, ¿estoy obteniendo los datos que debería? ¿o hay datos que parecen estar "sucios" o "incorrectos"?


In [ ]:
#Con la función isna y sum podemos obtener el número de nulos en cada columna
df.isna().sum()

producto           0
presentacion       0
marca              0
categoria          0
grupo              0
precio             0
fecha              0
marca_tienda       0
tipo_tienda        0
nombre_tienda      0
direccion          0
estado             0
municipio          0
latitud          907
longitud         907
dtype: int64

- La base de datos parece tener datos limpios, solo tenemos falta de datos en las columnas de latitud y longitud.
    

# POSTWORK 5 - LIMPIEZA DE DATOS

Este paso es requerido para poder utilizar apropiadamente el dataframe y conocer el contenido de las columnas. También vamos a explorar nuestro dataset usando agregaciones para ver si podemos empezar a responder algunas de las preguntas que nos hemos planteado. Los pasos que debes de seguir son los siguientes:

* Explora tu dataset con el fin de encontrar los NaNs que contiene. Piensa en la distribución de NaNs por columna y por fila.
* Piensa cuáles son los procedimientos que puedes aplicar a tus NaNs. ¿Tenemos que eliminar las filas/columnas que tienen esos NaNs? ¿O podríamos rellenar esos NaNs con algún valor de manera que podamos retener esas filas/columnas?
* Limpia tu dataset de manera que no quede ningún NaN.
* Reindexa tu dataset si lo consideras necesario.
* Renombra tus columnas si lo consideras necesario.
* Prueba aplicar agregaciones a tu DataFrame para ver si puedes empezar a responder algunas de las preguntas que te planteaste anteriormente. Aquí tienes una lista de algunas de las funciones que puedes utilizar para agregar/reducir tu dataset.

* Exploramos el dataset con el fin de encontrar los NaNs que contiene.

In [ ]:
#Se revisa la cantidad de filas y columnas, asi como su tipo de datos
print('Cantidad de Filas y columnas: ', df.shape)
print('Nombre columnas: ', df.dtypes)

Cantidad de Filas y columnas:  (256423, 15)
Nombre columnas:  producto          object
presentacion      object
marca             object
categoria         object
grupo             object
precio           float64
fecha             object
marca_tienda      object
tipo_tienda       object
nombre_tienda     object
direccion         object
estado            object
municipio         object
latitud          float64
longitud         float64
dtype: object


In [ ]:
#Se revisa la informacion del dataset y se detecta que las columnas latitud y longitud 
#tienen menos datos que las otras columnas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 256423 entries, 0 to 256422
Data columns (total 15 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   producto       256423 non-null  object 
 1   presentacion   256423 non-null  object 
 2   marca          256423 non-null  object 
 3   categoria      256423 non-null  object 
 4   grupo          256423 non-null  object 
 5   precio         256423 non-null  float64
 6   fecha          256423 non-null  object 
 7   marca_tienda   256423 non-null  object 
 8   tipo_tienda    256423 non-null  object 
 9   nombre_tienda  256423 non-null  object 
 10  direccion      256423 non-null  object 
 11  estado         256423 non-null  object 
 12  municipio      256423 non-null  object 
 13  latitud        255516 non-null  float64
 14  longitud       255516 non-null  float64
dtypes: float64(3), object(12)
memory usage: 29.3+ MB


In [ ]:
#Se revisa los valores nulos en las columnas y se confirma que existen valores nulos en las columnas latitud y longitud
df.isna().sum(axis=0)

producto           0
presentacion       0
marca              0
categoria          0
grupo              0
precio             0
fecha              0
marca_tienda       0
tipo_tienda        0
nombre_tienda      0
direccion          0
estado             0
municipio          0
latitud          907
longitud         907
dtype: int64

* Limpiamos el dataset eliminando los registros con valores NaN en latitud y longitud, debido a que varias de nuestras preguntas involucran estos datos por lo que estarían omitiéndose al realizar las consultas y la cantidad de registros con valores nulos representa sólo 0.003% del total.

In [ ]:
#Eliminar filas NaNs y filas duplicadas  
df = df.dropna(axis=0, how='any')
df = df.drop_duplicates()

* Asignamos un indice.

In [ ]:
#Se asigna un indice
df.insert(0, 'indice', range(1, 1 + len(df)))
df = df.set_index('indice')

df.head(2)

producto             presentacion  marca  \
indice                                            
1        ACEITE    BOTELLA 1 LT. VEGETAL  1-2-3   
2        ACEITE  BOTELLA 850 ML. VEGETAL    AVE   

                                categoria    grupo  precio       fecha  \
indice                                                                   
1       ACEITES Y GRASAS VEG. COMESTIBLES  BASICOS    56.0  2022-12-26   
2       ACEITES Y GRASAS VEG. COMESTIBLES    PACIC    43.9  2022-12-26   

        marca_tienda                            tipo_tienda  \
indice                                                        
1       MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
2       MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   

                               nombre_tienda  \
indice                                         
1       MEGA SORIANA SUCURSAL AGUASCALIENTES   
2       MEGA SORIANA SUCURSAL AGUASCALIENTES   

                                                direccion          estado  \
indice                                                                      
1       AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
2       AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   

             municipio    latitud    longitud  
indice                                         
1       AGUASCALIENTES  21.916001 -102.290779  
2       AGUASCALIENTES  21.916001 -102.290779

* Realizamos algunas agregaciones para conocer más sobre el contenido de los datos

In [ ]:
#verificar los análisis estadísticos de todas la columnas del dataframe
#Nos dice que hay 708 productos y 134 marcas de tienda distribuidos en 255516 filas
#el producto y marca de tienda que mas se repiten son REFRESCO y WAL-MART
df.describe(include = "all")

producto  presentacion   marca     categoria    grupo         precio  \
count     255516        255516  255516        255516   255516  255516.000000   
unique       708          2768     807            40        9            NaN   
top     REFRESCO  1 KG. GRANEL     S/M  MEDICAMENTOS  BASICOS            NaN   
freq        5338          3573   84827         54836   123377            NaN   
mean         NaN           NaN     NaN           NaN      NaN     632.072448   
std          NaN           NaN     NaN           NaN      NaN    2349.578727   
min          NaN           NaN     NaN           NaN      NaN       1.300000   
25%          NaN           NaN     NaN           NaN      NaN      29.500000   
50%          NaN           NaN     NaN           NaN      NaN      63.000000   
75%          NaN           NaN     NaN           NaN      NaN     315.600000   
max          NaN           NaN     NaN           NaN      NaN   67149.000000   

             fecha marca_tienda                            tipo_tienda  \
count       255516       255516                                 255516   
unique           5          134                                     13   
top     2022-12-27     WAL-MART  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
freq         59334        41517                                 205627   
mean           NaN          NaN                                    NaN   
std            NaN          NaN                                    NaN   
min            NaN          NaN                                    NaN   
25%            NaN          NaN                                    NaN   
50%            NaN          NaN                                    NaN   
75%            NaN          NaN                                    NaN   
max            NaN          NaN                                    NaN   

       nombre_tienda                                          direccion  \
count         255516                                             255516   
unique          1208                                               1266   
top         WAL MART  REVOLUCION 780, ENTRE RUBENS Y HOLBEIN, COL. M...   
freq            2371                                               1464   
mean             NaN                                                NaN   
std              NaN                                                NaN   
min              NaN                                                NaN   
25%              NaN                                                NaN   
50%              NaN                                                NaN   
75%              NaN                                                NaN   
max              NaN                                                NaN   

                  estado municipio        latitud       longitud  
count             255516    255516  255516.000000  255516.000000  
unique                29        67            NaN            NaN  
top     CIUDAD DE MÉXICO      LEÓN            NaN            NaN  
freq               56128     13850            NaN            NaN  
mean                 NaN       NaN      20.881830     -99.508514  
std                  NaN       NaN       2.936894       4.970139  
min                  NaN       NaN      16.736944    -117.041677  
25%                  NaN       NaN      19.345444    -101.647188  
50%                  NaN       NaN      19.614073     -99.199154  
75%                  NaN       NaN      21.153407     -99.005579  
max                  NaN       NaN      32.535127     -86.824104

In [ ]:
#Extraer codigo postal de la columna direccion
df['CP'] = df['direccion'].str[-5:]

In [ ]:
#Cantidad de códigos postales (509) 
df[['CP']].drop_duplicates().count()

CP    509
dtype: int64

* Investigamos sobre el contenido de la columna producto

In [ ]:
#verificar los análisis estadísticos de la columna producto
df['producto'].describe()

count       255516
unique         708
top       REFRESCO
freq          5338
Name: producto, dtype: object

In [ ]:
#Se realiza un conteo de los productos que se repiten mas
df.groupby("producto").size().sort_values(ascending=False)

producto
REFRESCO           5338
VARIOS             5264
JUEGOS DE MESA     4324
TOALLA FEMENINA    4083
PASTA PARA SOPA    3947
                   ... 
HALCION               1
CARPA                 1
JAIBON                1
COMPETACT             1
PARGO                 1
Length: 708, dtype: int64

Realizamos agregaciones para conocer más acerca de los valores en las columnas del dataset



In [ ]:
# Encontramos cuantos valores únicos hay en otras columnas de interés. 
# Así observamos que, por ejemplo, el dataset contiene 40 categorías de productos,
# hay 807 marcas de productos, etc.

df.nunique()

producto          708
presentacion     2768
marca             807
categoria          40
grupo               9
precio           9670
fecha               5
marca_tienda      134
tipo_tienda        13
nombre_tienda    1208
direccion        1266
estado             29
municipio          67
latitud          1153
longitud         1152
CP                509
dtype: int64

Realizamos agregaciones para conocer más acerca de las categorías de productos

In [ ]:
#Encontramos cuantos productos hay por categoría
df[['producto', 'categoria']].drop_duplicates().groupby('categoria').size().sort_values(ascending=False)

categoria
MEDICAMENTOS                          368
HORTALIZAS FRESCAS                     28
ARTS. PARA EL CUIDADO PERSONAL         28
PESCADOS Y MARISCOS                    25
FRUTAS FRESCAS                         24
PRODUCTOS DE TEMPORADA (NAVIDEÑOS)     20
APARATOS ELECTRICOS                    19
CONDIMENTOS                            18
ARTS. DE ESPARCIMIENTO (JUGUETES)      16
DERIVADOS DE LECHE                     15
FRUTAS Y LEGUMBRES PROCESADAS          15
CHOCOLATES Y GOLOSINAS                 14
APARATOS ELECTRONICOS                  14
DETERGENTES Y PRODUCTOS SIMILARES      11
CARNES FRIAS SECAS Y EMBUTIDOS          9
LEGUMBRES SECAS                         9
ARTS. DE PAPEL P/HIGIENE PERSONAL       8
ARROZ Y CEREALES PREPARADOS             7
REFRESCOS ENVASADOS                     7
GALLETAS PASTAS Y HARINAS DE TRIGO      7
PAN                                     7
LECHE PROCESADA                         6
VINOS Y LICORES                         6
TORTILLAS Y DERIVADOS DE

In [ ]:
df[['producto', 'categoria']].drop_duplicates().groupby('categoria').size().sort_values(ascending=False)

categoria
MEDICAMENTOS                          368
HORTALIZAS FRESCAS                     28
ARTS. PARA EL CUIDADO PERSONAL         28
PESCADOS Y MARISCOS                    25
FRUTAS FRESCAS                         24
PRODUCTOS DE TEMPORADA (NAVIDEÑOS)     20
APARATOS ELECTRICOS                    19
CONDIMENTOS                            18
ARTS. DE ESPARCIMIENTO (JUGUETES)      16
DERIVADOS DE LECHE                     15
FRUTAS Y LEGUMBRES PROCESADAS          15
CHOCOLATES Y GOLOSINAS                 14
APARATOS ELECTRONICOS                  14
DETERGENTES Y PRODUCTOS SIMILARES      11
CARNES FRIAS SECAS Y EMBUTIDOS          9
LEGUMBRES SECAS                         9
ARTS. DE PAPEL P/HIGIENE PERSONAL       8
ARROZ Y CEREALES PREPARADOS             7
REFRESCOS ENVASADOS                     7
GALLETAS PASTAS Y HARINAS DE TRIGO      7
PAN                                     7
LECHE PROCESADA                         6
VINOS Y LICORES                         6
TORTILLAS Y DERIVADOS DE

In [ ]:
#Encontramos cuantos tiendas hay por estado, agrupando por direccion de la tienda y estado
df[['estado', 'direccion']].drop_duplicates().groupby('estado').size().sort_values(ascending=False)

estado
CIUDAD DE MÉXICO        192
ESTADO DE MÉXICO        152
JALISCO                  72
GUANAJUATO               68
ZACATECAS                52
QUINTANA ROO             52
VERACRUZ                 51
NUEVO LEÓN               48
QUERÉTARO                44
YUCATÁN                  41
TABASCO                  38
OAXACA                   37
SINALOA                  35
PUEBLA                   32
TLAXCALA                 28
MORELOS                  27
DURANGO                  27
COAHUILA DE ZARAGOZA     27
CHIAPAS                  27
AGUASCALIENTES           26
MICHOACÁN DE OCAMPO      25
CHIHUAHUA                25
BAJA CALIFORNIA          24
TAMAULIPAS               24
CAMPECHE                 24
BAJA CALIFORNIA SUR      24
SAN LUIS POTOSÍ          21
SONORA                   20
GUERRERO                  3
dtype: int64

De acuerdo con el conocimiento obtenido sobre el contenido del dataset al parecer si es posible encontrar los precios menores de productos y categorías usando agregaciones, pero se requiere crear un dato adicional que nos permita homogenizar los precios por unidad de medida, por ejemplo, precio por gramo o kilogramo, porque existen diferentes presentaciones de un mismo producto.
También técnicamente es posible proporcionar las tiendas cercanas a una ubicación que ofrezcan cierto producto con menor precio porque los datos muestran precisión, sin embargo, la cantidad de códigos postales (509) es muy pequeña respecto al total de códigos postales en México (30 mil aproximadamente), por lo que estos datos al parecer son muestrales.

# POSTWORK 6 - APIS, AUTOMATIZACIÓN Y CONCATENACIÓN DE DATAFRAMES

En este postwork buscamos un API que complementara nuestro dataset original. Aunque PROFECO si mantiene publicada un API sobre "Quién es quien en los precios" y está en operación, al solicitarle datos regresa una estructura vacía y no encontramos documentación acerca de los parámetros apropiados para realizar las solicitudes.

Buscamos un API que estuviera relacionada con nuestro dataset original y encontramos interesante explorar el [API de COPOMEX](https://api.copomex.com/documentacion/inicio) (Código Postal Mexicano), dedicada a la ubicación, búsqueda y manejo de códigos postales en México.

Creamos una cuenta, exploramos la documentación y realizamos peticiones de prueba con éxito.

[Ver Jupyter Notebook](https://github.com/adavals/bedu-datascience-f2m2/blob/main/postwork6.ipynb)



# POSTWORK 7 - TRANSFORMACIÓN, FILTRACIÓN Y REORDENAMIENTO

Para el correcto análisis y explotación de la información la mayoría de las veces es necesario aplicar algunas transformaciones que incluyen el filtrado para conservar únicamente la información objeto de nuestro estudio, la creación de columnas calculadas para un propósito específico, el recasteo de columnas con tipos de datos incorrectos y el agrupamiento y reordenamiento de la información que en ocasiones derivan en nuevos datasets.

* Tras el análisis exploratorio de datos nos percatamos de que las fechas no se encuentran en el formato correcto, por lo que es necesario aplicar un recasteo.

In [ ]:
#Casting de fechas a datetime64[ns]
df['fecha']=pd.to_datetime(df['fecha'], unit='ns')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 255516 entries, 1 to 255516
Data columns (total 16 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   producto       255516 non-null  object        
 1   presentacion   255516 non-null  object        
 2   marca          255516 non-null  object        
 3   categoria      255516 non-null  object        
 4   grupo          255516 non-null  object        
 5   precio         255516 non-null  float64       
 6   fecha          255516 non-null  datetime64[ns]
 7   marca_tienda   255516 non-null  object        
 8   tipo_tienda    255516 non-null  object        
 9   nombre_tienda  255516 non-null  object        
 10  direccion      255516 non-null  object        
 11  estado         255516 non-null  object        
 12  municipio      255516 non-null  object        
 13  latitud        255516 non-null  float64       
 14  longitud       255516 non-null  float64       
 15  

* Por cuestiones prácticas nos concentraremos en los productos del catálogo PACIC (Paquete Contra la Inflación y la Carestía) que se enfoca en productos de la canasta básica, y dentro de este grupo, solo a ciertas categorías de productos.

In [ ]:
#Filtrado por grupo PACIC y productos de la canasta básica
grupo_pacic = df['grupo']=='PACIC'
df=df[grupo_pacic]

df = df[df.categoria.isin(['ACEITES Y GRASAS VEG. COMESTIBLES','ARROZ Y CEREALES PREPARADOS','AZUCAR','FRUTAS FRESCAS',
'HORTALIZAS FRESCAS','LECHE FRESCA','TORTILLAS Y DERIVADOS DEL MAIZ'])]

df.head(3)

producto                        presentacion       marca  \
indice                                                            
2        ACEITE             BOTELLA 850 ML. VEGETAL         AVE   
29        ARROZ  BOLSA 900 GR. EXTRA. (GRANO LARGO)  PRECISSIMO   
47       AZUCAR     BOLSA PLÁSTICO 900 GR. ESTÁNDAR  PRECISSIMO   

                                categoria  grupo  precio      fecha  \
indice                                                                
2       ACEITES Y GRASAS VEG. COMESTIBLES  PACIC    43.9 2022-12-26   
29            ARROZ Y CEREALES PREPARADOS  PACIC    16.9 2022-12-26   
47                                 AZUCAR  PACIC    19.9 2022-12-26   

        marca_tienda                            tipo_tienda  \
indice                                                        
2       MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
29      MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
47      MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   

                               nombre_tienda  \
indice                                         
2       MEGA SORIANA SUCURSAL AGUASCALIENTES   
29      MEGA SORIANA SUCURSAL AGUASCALIENTES   
47      MEGA SORIANA SUCURSAL AGUASCALIENTES   

                                                direccion          estado  \
indice                                                                      
2       AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
29      AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
47      AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   

             municipio    latitud    longitud     CP  
indice                                                
2       AGUASCALIENTES  21.916001 -102.290779  20116  
29      AGUASCALIENTES  21.916001 -102.290779  20116  
47      AGUASCALIENTES  21.916001 -102.290779  20116

* Con la finalidad de comparar los precios de los distintos productos de manera consistente es necesario homogenizar las unidades de medida ya que se venden en diferentes presentaciones, para ello primero obtendremos el contenido en gramos y mililitros y calcularemos su precio por kilogramo o litro según corresponda.

In [ ]:
#Extraer cantidad en g y ml de presentación
def extraer_numeros(s):
    s=[int(temp)for temp in s.split() if temp.isdigit()]
    return s[0]

df['uom_std']=df['presentacion'].map(extraer_numeros)

#Calcular precio por kilogramo o litro
def precio_x_uom(uom, precio):
      if uom == 1:
        precio_std = precio
      else:
        precio_std = precio*1000/uom
      return round(precio_std,2)
      
df['precio_kg_l']=df.apply(lambda x: precio_x_uom(x['uom_std'],x['precio']), axis=1)

df.head(5)

producto                                      presentacion  \
indice                                                                   
2             ACEITE                           BOTELLA 850 ML. VEGETAL   
29             ARROZ                BOLSA 900 GR. EXTRA. (GRANO LARGO)   
47            AZUCAR                   BOLSA PLÁSTICO 900 GR. ESTÁNDAR   
120     CHILE FRESCO                       1 KG. JALAPEÑO O CUARESMEÑO   
301         JITOMATE  1 KG. SALADETTE/ HUAJE O TOMATE SALADETTE/ HUAJE   

             marca                          categoria  grupo  precio  \
indice                                                                 
2              AVE  ACEITES Y GRASAS VEG. COMESTIBLES  PACIC    43.9   
29      PRECISSIMO        ARROZ Y CEREALES PREPARADOS  PACIC    16.9   
47      PRECISSIMO                             AZUCAR  PACIC    19.9   
120            S/M                 HORTALIZAS FRESCAS  PACIC    28.9   
301            S/M                 HORTALIZAS FRESCAS  PACIC    19.8   

            fecha  marca_tienda                            tipo_tienda  \
indice                                                                   
2      2022-12-26  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
29     2022-12-26  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
47     2022-12-26  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
120    2022-12-26  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   
301    2022-12-26  MEGA SORIANA  SUPERMERCADO / TIENDA DE AUTOSERVICIO   

                               nombre_tienda  \
indice                                         
2       MEGA SORIANA SUCURSAL AGUASCALIENTES   
29      MEGA SORIANA SUCURSAL AGUASCALIENTES   
47      MEGA SORIANA SUCURSAL AGUASCALIENTES   
120     MEGA SORIANA SUCURSAL AGUASCALIENTES   
301     MEGA SORIANA SUCURSAL AGUASCALIENTES   

                                                direccion          estado  \
indice                                                                      
2       AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
29      AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
47      AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
120     AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   
301     AGUASCALIENTES NTE. 802, ESQ. BLVD. A ZACATECA...  AGUASCALIENTES   

             municipio    latitud    longitud     CP  uom_std  precio_kg_l  
indice                                                                      
2       AGUASCALIENTES  21.916001 -102.290779  20116      850        51.65  
29      AGUASCALIENTES  21.916001 -102.290779  20116      900        18.78  
47      AGUASCALIENTES  21.916001 -102.290779  20116      900        22.11  
120     AGUASCALIENTES  21.916001 -102.290779  20116        1        28.90  
301     AGUASCALIENTES  21.916001 -102.290779  20116        1        19.80

# POSTWORK 8 - PUBLICACION Y TRABAJOS FUTUROS

El trabajo realizado y recopilado en este Jupyter Notebook también ha sido publicado en el siguiente repositorio GitHub: https://github.com/adavals/bedu-datascience-f2m2.git 





# TRABAJOS FUTUROS
Considerando la utilidad para el consumidor y las preguntas que nos planteamos al inicio será interesante:
- Encontrar precios promedio para productos, categorias, de acuerdo al area geográfica, como una referencia para el consumidor.
- Analizar y segmentar el nivel (bajo, medio, alto) de  precios por areas geograficas.
- Encontrar la lista de tiendas alrededor de una ubicación en particular.
- Representar en forma gráfica los análisis y consultas anteriores.
- Considerando varios dataset a través del tiempo, realizar predicciones de precios.
### Más a futuro sería interesante:
* Investigar fuentes adicionales de precios del día o en tiempo real para un mayor numero de tiendas, ya sea con APIs existentes de los centros comerciales o web scraping para generar dataset con más cobertura, no sólo la muestra de PROFECO.
* O, podría desarrollarse un API para comerciantes donde ellos puedan alimentar el dataset con información del día sobre los precios de los productos y una aplicación para el consumidor donde se le alerte de precios bajos de su interés en su área.
* Con la información de precios y de perfil de consultas de los consumidores se pueden encontrar patrones de consulta/consumo y de fluctuaciones en los precios. 


Investigamos un poco sobre cómo determinar la distancia en km entre ubicaciones geográficas mediante latitud y longitud

In [ ]:
!pip install mpu


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.7/69.7 KB 1.1 MB/s eta 0:00:00


In [ ]:
import mpu

In [ ]:
munich = (48.1372, 11.5756)
berlin = (52.5186, 13.4083)
round(mpu.haversine_distance(munich, berlin), 1)


504.2